### DISCUSSION TITLE
swanctl output in JSON

### DISCUSSION THREAD

**Discussion by misch42** [strongswan/strongswan#1841] [Ideas]
Hi,

data processing systems like to exchange data in defined data formats. i.e.  in INT or STRING.

modern systems more and more exchange data in JSON format.

It would be very nice to have the option --json in the swanctl command. So the output of swanctl --list-sas or --list-conns would be JSON directly.

Even better would be if the swanctl command would accept configurations or config data directly in JSON. So strongswan could be easily configured from config management systems like ansible without the need to write jinja templates.

Michael.

**Comment by strongX509** [comment]
The strongSwan **`vici`** interface uses a binary encoding of a data format that is very close to the JSON format. It can be printed with the
**`swanctl --list-sas --raw [--pretty]`** option: 
```
list-sa event {
  rw {
    uniqueid = 567
    version = 2
    state = ESTABLISHED
    local-host = 10.12.1.1
    local-port = 500
    local-id = moon.strongswan.org
    remote-host = 10.12.1.17
    remote-port = 500
    remote-id = carol@strongswan.org
    initiator-spi = 5b13b329c572709d
    responder-spi = e8fba447049c3cbf
    encr-alg = AES_CBC
    encr-keysize = 256
    integ-alg = HMAC_SHA2_384_192
    prf-alg = PRF_HMAC_SHA2_384
    dh-group = ECP_384
    established = 5509
    reauth-time = 21380
    child-sas {
      dbmas-1481 {
        name = dbmas
        uniqueid = 1481
        reqid = 3
        state = INSTALLED
        mode = TUNNEL
        protocol = ESP
        spi-in = c24c6cbe
        spi-out = c0d4ac53
        encr-alg = AES_GCM_16
        encr-keysize = 256
        bytes-in = 119472
        packets-in = 540
        use-in = 28
        bytes-out = 129468
        packets-out = 540
        use-out = 28
        rekey-time = 7788
        life-time = 10331
        install-time = 5509
        local-ts = [
          10.12.0.0/24
        ]
        remote-ts = [
          10.12.1.17/32
        ]
      }
    }
  }
}
```

**Reply by misch42** [reply]
thanks for the fast reply.

I know about the "pretty|raw" option. Output close to JSON, but not still other scripts cannot work with this data without heavy parsing. So I asked for an output in genuine JSON. So it would be much easier to work with the data further.

Or the other way: Configure strongswan via vici in plain JSON via config management like ansible or saltstack.

**Reply by strongX509** [reply]
Well, since the **`swanctl`** command line tool also parses **`swanctl.conf`** which isn't written in pure JSON syntax either, an additional **`--json`** option wouldn't make much sense. 

Thus a separate JSON-compliant tool would be required for converting inputs and outputs between the *VICI* binary interface format and UTF-8 based JSON format. This tool could be either written in C using **`libvici`** (requiring the JSON converter to be released under the GPLv2 license) or in one of the *Python*, *Ruby*, *Perl* or *Go* script languages for which *VICI* interface wrappers already exist (allowing the new tool to be released under any license).

**Comment by misch42** [comment]
Yes. I think, that would be the was to go. Thanks.

**Reply by syzdek** [reply]
I am working on a project, [davici-utils](https://github.com/syzdek/davici-utils), which uses the davici library to duplicates some of the functionality of swanctl.  I currently have most of the vici commands working which query the daemon and utility returns the results in json, xml, or yaml.

**Comment by tobiasbrunner** [comment]
This is also a duplicate of #493.

**Comment by plan9better** [comment]
If you also have this issue, there are some solutions [here](https://github.com/strongswan/strongswan/issues/493#issuecomment-2980965241)
